### PR TITLE
Further improvements for polygonize command

### PR DESCRIPTION
**Pull request by m-mohr**
- Allows to select file format based on file extension given in --out parameter (Geopackage, GeoParquet [no fiboa yet, has vertain GDAL requirements], GeoJSON, FlatGeoBuf)
- Restructured code to have some central information in cfg.py
- close_interiors option to remove holes from polygons
- Renamed property idx to id in output features (fiboa alignment), start with 1 instead of 0, string instead of int

Based on #50 , which needs to be merged first.

### CODE REVIEW

**Comment by cholmes** [comment]
> GeoParquet [no fiboa yet, has vertain GDAL requirements]

What are the gdal requirements? Just the arrow libs gdal extension? We may consider using geopandas, as I believe it has an implementation of [GeoParquet output](https://geopandas.org/en/stable/docs/reference/api/geopandas.GeoDataFrame.to_parquet.html) that does not rely on GDAL.

**Comment by m-mohr** [comment]
Gdal 3.8+ and the lib arrow requirements, yes.

Once I implement fiboa I'll reuse fiboa cli, which uses pyarrow, indeed. So for Parquet this solution is temporary.

File: src/ftw_cli/cli.py [hunk 1]
```diff
@@ -106,13 +108,14 @@ def inference_run(input, model, out, resize_factor, gpu, patch_size, batch_size,
 
 @inference.command("polygonize", help="Polygonize the output from inference for the raster image given via INPUT.")
 @click.argument('input', type=click.Path(exists=True), required=True)
-@click.option('--out', '-o', type=str, required=True, help="Output filename for the polygonized data.")
+@click.option('--out', '-o', type=str, required=True, help="Output filename for the polygonized data. " + SUPPORTED_POLY_FORMATS_TXT)
 @click.option('--simplify', type=float, default=None, help="Simplification factor to use when polygonizing.")
 @click.option('--min_size', type=float, default=500, help="Minimum area size in square meters to include in the output.")
```
Comment (cholmes) [inline src/ftw_cli/cli.py:113]:
```suggestion
@click.option('--min_size', type=float, default=500, help="Minimum area size in square meters to include in the output (default is 500).")
```

File: src/ftw_cli/cli.py [hunk 2]
```diff
@@ -106,13 +108,14 @@ def inference_run(input, model, out, resize_factor, gpu, patch_size, batch_size,
 
 @inference.command("polygonize", help="Polygonize the output from inference for the raster image given via INPUT.")
 @click.argument('input', type=click.Path(exists=True), required=True)
-@click.option('--out', '-o', type=str, required=True, help="Output filename for the polygonized data.")
+@click.option('--out', '-o', type=str, required=True, help="Output filename for the polygonized data. " + SUPPORTED_POLY_FORMATS_TXT)
 @click.option('--simplify', type=float, default=None, help="Simplification factor to use when polygonizing.")
```
Comment (cholmes) [inline src/ftw_cli/cli.py:112]:
```suggestion
@click.option('--simplify', type=float, default=15, help="Simplification factor to use when polygonizing. Assumes meters, default is 15.")
```

File: src/ftw_cli/cli.py [hunk 3]
```diff
@@ -106,13 +108,14 @@ def inference_run(input, model, out, resize_factor, gpu, patch_size, batch_size,
...
 @click.option('--min_size', type=float, default=500, help="Minimum area size in square meters to include in the output.")
 @click.option('--overwrite', '-f', is_flag=True, help="Overwrite outputs if they exist.")
-def inference_polygonize(input, out, simplify, min_size, overwrite):
+@click.option('--close_interiors', is_flag=True, help="Remove the interiors holes in the polygons.")
```
Comment (cholmes) [inline src/ftw_cli/cli.py:115]: This is a great start. I do think we should figure out how to only holes below X size, and then remove this as an option (and we maybe just use simplify to mean remove both small polys and small holes). We probably don't want to default it to true in this state

**Review by cholmes — COMMENTED** [review]
Hrm, just testing this and it seems to have stopped working right. It starts up, but hangs on 'processing mask windows: 0%'. Hitting ctrl-c on the command-line many times doesn't do anything. Killing the python process from mac 'activity monitor' works and prints: 
```
/Users/cholmes/miniforge3/envs/ftw-c2/lib/python3.12/multiprocessing/resource_tracker.py:254: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
```

I'll investigate more later today. Other than that looks great, put in a few comments / suggestions. Curious if there's any reason you didn't also do the -c style short cuts on these (I tend not to use those, so no prob for me, but we have it on -o).

**Comment by cholmes** [comment]
Oh wait, I just let it run, and it looks like it's not actually hanging. It's just taking a long time and not showing progress:

```
ftw inference polygonize nl_output.tif -o dude5.gpkg                   
/Users/cholmes/miniforge3/envs/ftw-c2/lib/python3.12/site-packages/kornia/feature/lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
Polygonizing input file: nl_output.tif
Processing mask windows:   0%|                                                                | 0/1 [00:00<?, ?it/s]

Processing mask windows: 100%|███████████████████████████████████████████████████████| 1/1 [06:22<00:00, 382.68s/it]
Finished polygonizing output at dude5.gpkg in 383.26s
```

File: src/ftw_cli/cli.py (hunk 1 above)
Comment (m-mohr) [reply src/ftw_cli/cli.py:113]: The better thing to do is:

```suggestion
@click.option('--min_size', type=float, default=500, show_default=True, help="Minimum area size in square meters to include in the output.")
```

File: src/ftw_cli/cli.py (hunk 2 above)
Comment (m-mohr) [reply src/ftw_cli/cli.py:112]: Is this actually always meters or does it depend on the CRS unit?
Do we have any evidence on the 15?

File: src/ftw_cli/cli.py (hunk 1 above)
Comment (cholmes) [reply src/ftw_cli/cli.py:113]: Ah, perfect - as I wrote that I felt there was likely a better way but didn't see other examples right there.

File: src/ftw_cli/cli.py (hunk 3 above)
Comment (m-mohr) [reply src/ftw_cli/cli.py:115]: See #39 for discussions.

**Comment by m-mohr** [comment]
Hmm, weird, but I also saw this behaviour before this PR on my side. tbc

File: src/ftw_cli/cli.py (hunk 2 above)
Comment (cholmes) [reply src/ftw_cli/cli.py:112]: > Is this actually always meters or does it depend on the CRS unit?

Depends on the CRS unit. But we only have S2 as input right now. So I lean towards assuming that in our code - but would be better for now if we raised an error if the projection was not in meters, and say that's not supported. Vs having the user have to understand the nature of their CRS to supply a simplify parameter. I guess we also could just translate meters to other CRS's. I'd very much like this call to assume meters - I don't like the simplify calls in the geo world that force you to know / look up what units your CRS is in.

File: src/ftw_cli/inference.py [hunk 4]
```diff
@@ -235,56 +233,80 @@ def polygonize(input, out, simplify, min_size, overwrite):
...
     # read the input file as a mask
     with rasterio.open(input) as src:
         input_height, input_width = src.shape
         original_crs = src.crs.to_string()
+        is_meters = src.crs.linear_units in ["m", "metre", "meter"]
         transform = src.transform 
         mask = (src.read(1) == 1).astype(np.uint8)
-        polygonization_stride = 2048
-        total_iterations = (input_height // polygonization_stride) * (input_width // polygonization_stride)
+        polygonization_stride = 10980
```
Comment (m-mohr) [inline src/ftw_cli/inference.py:245]:
```suggestion
        polygonization_stride = 2048
```

**Comment by m-mohr** [comment]
@cholmes I accidentally comitted a different polygonization_stride size, which lead to the issue. Reverted back to 2048, so it should work better again.

File: src/ftw_cli/cli.py (hunk 2 above)
Comment (m-mohr) [reply src/ftw_cli/cli.py:112]: The area computation has been improved to work on any CRS. If we restrict to m, we could potentially also remove that code.

On the other hand, I feel like it's fine to just elaborate in the description. I mean it's still meters in many cases, including S2. Translating meters to other CRSes is not that easy.

**Comment by cholmes** [comment]
Tested, much better:
```
Polygonizing input file: nl_output.tif
Processing mask windows: 100%|██████████████████████████████████████████████████████| 36/36 [00:34<00:00,  1.04it/s]
Finished polygonizing output at dude7.fgb in 35.25s
```

File: src/ftw_cli/cli.py (hunk 2 above)
Comment (m-mohr) [reply src/ftw_cli/cli.py]: See 2fa233ed

**Review by cholmes — APPROVED** [review]

**Pull request merged**